File: AImmoold/frontend/lib/types/rag.ts
import { SourceType } from "./document"

/**
 * RAG Types - Phase 4 Foundation
 * Chunking, vectorisation, indexation et recherche
 */

// Chunk Status
export enum ChunkStatus {
    PENDING = "pending",
    PROCESSING = "processing",
    INDEXED = "indexed",
    FAILED = "failed",
    EXCLUDED = "excluded",
}

// Chunk représentant un fragment de document
export interface Chunk {
    id: string
    document_id: string
    lease_id?: string
    property_id?: string
    organization_id: string

    // Contenu
    content: string
    content_hash: string

    // Métadonnées de chunking
    chunk_index: number
    total_chunks: number
    start_offset: number
    end_offset: number

    // Source et type
    source_type: SourceType
    source_id: string // ID du document/bail/propriété source

    // Vectorisation
    embedding?: number[]
    embedding_model?: string

    // Tags sémantiques automatiques
    semantic_tags: string[]

    // Contrôle RAG
    status: ChunkStatus
    is_excluded: boolean // Exclusion manuelle du RAG

    // Métadonnées
    metadata: ChunkMetadata
    created_at: string
    updated_at: string
}

// Métadonnées enrichies du chunk
export interface ChunkMetadata {
    // Titre du document source
    source_title: string
    // Type de document (bail, facture, etc.)
    document_type?: string
    // Page (pour PDFs)
    page_number?: number
    // Section/heading détecté
    section_heading?: string
    // Langue détectée
    language?: string
    // Score de qualité du chunk
    quality_score?: number
    // Contexte additionnel
    context?: Record<string, unknown>
}

// Configuration de chunking par type de source
export interface ChunkingConfig {
    source_type: SourceType
    // Taille de chunk en tokens
    chunk_size: number
    // Chevauchement entre chunks
    chunk_overlap: number
    // Méthode de chunking
    chunking_method: ChunkingMethod
    // Séparateurs personnalisés
    separators?: string[]
}

export enum ChunkingMethod {
    // Découpage par taille fixe
    FIXED_SIZE = "fixed_size",
    // Découpage par paragraphes
    PARAGRAPH = "paragraph",
    // Découpage sémantique (titres, sections)
    SEMANTIC = "semantic",
    // Découpage par sentences
    SENTENCE = "sentence",
    // Récursif (essaie différents séparateurs)
    RECURSIVE = "recursive",
}

// Configuration RAG globale
export interface RAGConfig {
    organization_id: string

    // Modèle d'embedding
    embedding_model: EmbeddingModel

    // Configurations de chunking par source
    chunking_configs: Record<SourceType, ChunkingConfig>

    // Visibilité des sources dans les recherches
    source_visibility: Record<SourceType, boolean>

    // Documents exclus manuellement
    excluded_document_ids: string[]

    // Tags pour filtrer
    tag_filters: string[]

    // Paramètres de recherche
    search_config: SearchConfig
}

export enum EmbeddingModel {
    OPENAI_ADA = "text-embedding-ada-002",
    OPENAI_3_SMALL = "text-embedding-3-small",
    OPENAI_3_LARGE = "text-embedding-3-large",
    COHERE_MULTILINGUAL = "embed-multilingual-v3.0",
    VOYAGE_LARGE = "voyage-large-2-instruct",
}

// Configuration de recherche
export interface SearchConfig {
    // Nombre de chunks à retourner
    top_k: number
    // Score minimum de similarité
    min_similarity: number
    // Reranking activé
    use_reranking: boolean
    // Modèle de reranking
    reranking_model?: string
    // Expansion de requête
    use_query_expansion: boolean
}

// Requête de recherche RAG
export interface RAGSearchRequest {
    organization_id: string
    query: string
    // Filtres de source
    source_types?: SourceType[]
    // Filtres par document/bail/propriété
    document_ids?: string[]
    lease_ids?: string[]
    property_ids?: string[]
    // Filtres par tags
    tags?: string[]
    // Limite de résultats
    limit?: number
    // Score minimum
    min_score?: number
}

// Résultat de recherche RAG
export interface RAGSearchResult {
    chunk: Chunk
    score: number
    // Highlights pour affichage
    highlights?: string[]
    // Contexte étendu (chunks adjacents)
    context_before?: string
    context_after?: string
}

// Réponse de recherche RAG
export interface RAGSearchResponse {
    results: RAGSearchResult[]
    total_chunks_searched: number
    query_embedding?: number[]
    processing_time_ms: number
    // Sources utilisées
    sources_included: SourceType[]
}

// État d'indexation d'un document
export interface DocumentIndexStatus {
    document_id: string
    status: IndexStatus
    chunks_count: number
    chunks_indexed: number
    last_indexed_at?: string
    error_message?: string
}

export enum IndexStatus {
    NOT_INDEXED = "not_indexed",
    INDEXING = "indexing",
    INDEXED = "indexed",
    PARTIAL = "partial",
    FAILED = "failed",
    EXCLUDED = "excluded",
}

// Statistiques RAG par organisation
export interface RAGStats {
    organization_id: string
    total_chunks: number
    chunks_by_source: Record<SourceType, number>
    documents_indexed: number
    documents_excluded: number
    last_index_update: string
    storage_used_bytes: number
}

// Job d'indexation
export interface IndexingJob {
    id: string
    organization_id: string
    document_ids: string[]
    status: JobStatus
    progress: number
    started_at: string
    completed_at?: string
    error?: string
}

export enum JobStatus {
    PENDING = "pending",
    RUNNING = "running",
    COMPLETED = "completed",
    FAILED = "failed",
    CANCELLED = "cancelled",
}

// Toggle de source pour l'UI
export interface SourceToggle {
    source_type: SourceType
    label: string
    icon: string
    enabled: boolean
    count: number
}

// État du contexte RAG
export interface RAGContextState {
    config: RAGConfig | null
    stats: RAGStats | null
    sourceToggles: SourceToggle[]
    excludedDocuments: Set<string>
    isLoading: boolean
    error: string | null
}

File: AImmoold/backend/app/schemas/rag.py
"""
RAG Schemas - Phase 4 Foundation
Chunking, vectorisation, indexation et recherche
"""

from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional, List, Dict, Any
from uuid import UUID
from enum import Enum


class SourceType(str, Enum):
    """Types de sources pour l'indexation RAG"""
    DOCUMENT = "document"
    LEASE = "lease"
    PROPERTY = "property"
    TENANT = "tenant"
    KPI = "kpi"
    CONVERSATION = "conversation"


class ChunkStatus(str, Enum):
    """Statut d'un chunk"""
    PENDING = "pending"
    PROCESSING = "processing"
    INDEXED = "indexed"
    FAILED = "failed"
    EXCLUDED = "excluded"


class ChunkingMethod(str, Enum):
    """Méthodes de chunking"""
    FIXED_SIZE = "fixed_size"
    PARAGRAPH = "paragraph"
    SEMANTIC = "semantic"
    SENTENCE = "sentence"
    RECURSIVE = "recursive"


class EmbeddingModel(str, Enum):
    """Modèles d'embedding disponibles"""
    OPENAI_ADA = "text-embedding-ada-002"
    OPENAI_3_SMALL = "text-embedding-3-small"
    OPENAI_3_LARGE = "text-embedding-3-large"


class IndexStatus(str, Enum):
    """Statut d'indexation d'un document"""
    NOT_INDEXED = "not_indexed"
    INDEXING = "indexing"
    INDEXED = "indexed"
    PARTIAL = "partial"
    FAILED = "failed"
    EXCLUDED = "excluded"


# ============================================
# Chunk Models
# ============================================

class ChunkMetadata(BaseModel):
    """Métadonnées d'un chunk"""
    source_title: str = ""
    document_type: Optional[str] = None
    page_number: Optional[int] = None
    section_heading: Optional[str] = None
    language: str = "fr"
    quality_score: Optional[float] = None
    context: Optional[Dict[str, Any]] = None


class ChunkBase(BaseModel):
    """Base d'un chunk"""
    document_id: UUID
    organization_id: UUID
    content: str
    chunk_index: int
    total_chunks: int
    start_offset: int
    end_offset: int
    source_type: SourceType
    source_id: UUID
    semantic_tags: List[str] = []
    metadata: ChunkMetadata


class ChunkCreate(ChunkBase):
    """Création d'un chunk"""
    content_hash: str
    embedding: Optional[List[float]] = None
    embedding_model: Optional[str] = None


class Chunk(ChunkBase):
    """Chunk complet"""
    id: UUID
    content_hash: str
    embedding: Optional[List[float]] = None
    embedding_model: Optional[str] = None
    status: ChunkStatus = ChunkStatus.PENDING
    is_excluded: bool = False
    lease_id: Optional[UUID] = None
    property_id: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ============================================
# Chunking Configuration
# ============================================

class ChunkingConfig(BaseModel):
    """Configuration de chunking par type de source"""
    source_type: SourceType
    chunk_size: int = 512
    chunk_overlap: int = 50
    chunking_method: ChunkingMethod = ChunkingMethod.RECURSIVE
    separators: Optional[List[str]] = None


# ============================================
# Indexation Requests
# ============================================

class IndexDocumentRequest(BaseModel):
    """Requête d'indexation d'un document"""
    document_id: UUID
    content: str
    source_type: SourceType = SourceType.DOCUMENT
    metadata: Optional[ChunkMetadata] = None
    chunking_config: Optional[ChunkingConfig] = None


class IndexDocumentResponse(BaseModel):
    """Réponse d'indexation d'un document"""
    document_id: UUID
    status: IndexStatus
    chunks_count: int
    chunks_indexed: int
    last_indexed_at: Optional[datetime] = None
    error_message: Optional[str] = None


class BulkIndexRequest(BaseModel):
    """Requête d'indexation en masse"""
    document_ids: List[UUID]
    force_reindex: bool = False


class BulkIndexResponse(BaseModel):
    """Réponse d'indexation en masse"""
    total_documents: int
    indexed: int
    failed: int
    results: List[IndexDocumentResponse]


# ============================================
# Search Requests
# ============================================

class RAGSearchRequest(BaseModel):
    """Requête de recherche RAG"""
    organization_id: UUID
    query: str
    source_types: Optional[List[SourceType]] = None
    document_ids: Optional[List[UUID]] = None
    lease_ids: Optional[List[UUID]] = None
    property_ids: Optional[List[UUID]] = None
    tags: Optional[List[str]] = None
    limit: int = Field(default=10, ge=1, le=50)
    min_score: float = Field(default=0.5, ge=0, le=1)


class RAGSearchResult(BaseModel):
    """Résultat de recherche RAG"""
    chunk_id: UUID
    document_id: UUID
    content: str
    score: float
    source_type: SourceType
    metadata: ChunkMetadata
    semantic_tags: List[str]
    highlights: Optional[List[str]] = None


class RAGSearchResponse(BaseModel):
    """Réponse de recherche RAG"""
    results: List[RAGSearchResult]
    total_chunks_searched: int
    processing_time_ms: int
    sources_included: List[SourceType]


# ============================================
# RAG Configuration
# ============================================

class RAGConfigUpdate(BaseModel):
    """Mise à jour de la config RAG"""
    source_visibility: Optional[Dict[SourceType, bool]] = None
    excluded_document_ids: Optional[List[UUID]] = None
    tag_filters: Optional[List[str]] = None
    embedding_model: Optional[EmbeddingModel] = None


class RAGConfig(BaseModel):
    """Configuration RAG complète"""
    organization_id: UUID
    embedding_model: EmbeddingModel = EmbeddingModel.OPENAI_3_SMALL
    chunking_configs: Dict[SourceType, ChunkingConfig] = {}
    source_visibility: Dict[SourceType, bool] = {}
    excluded_document_ids: List[UUID] = []
    tag_filters: List[str] = []


# ============================================
# RAG Stats
# ============================================

class RAGStats(BaseModel):
    """Statistiques RAG"""
    organization_id: UUID
    total_chunks: int
    chunks_by_source: Dict[SourceType, int]
    documents_indexed: int
    documents_excluded: int
    last_index_update: Optional[datetime] = None
    storage_used_bytes: int = 0


# ============================================
# Document Exclusion
# ============================================

class DocumentExclusionRequest(BaseModel):
    """Requête d'exclusion/inclusion d'un document"""
    document_id: UUID
    excluded: bool


class DocumentExclusionResponse(BaseModel):
    """Réponse d'exclusion/inclusion"""
    document_id: UUID
    is_excluded: bool
    chunks_affected: int
